home-manager-auto-upgrade

s0racat opened the pull request:
### Description

- Added a new preSwitchCommands option to allow users to run arbitrary commands before home-manager switch.
- Added a new flags option to allow users to pass additional options to home-manager switch
- fix #8349 
<!--

Please provide a brief description of your change.

-->

### Checklist

<!--

Please go through the following checklist before opening a non-WIP
pull-request.

Also make sure to read the guidelines found at

  https://nix-community.github.io/home-manager/#sec-guidelines

-->

- [ ] Change is backwards compatible.

- [x] Code formatted with `nix fmt` or
      `nix-shell -p treefmt nixfmt deadnix keep-sorted --run treefmt`.

- [ ] Code tested through `nix run .#tests -- test-all` or
      `nix-shell --pure tests -A run.all`.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [] Commit messages are formatted like

  ```
  {component}: {description}

  {long description}
  ```

  See [CONTRIBUTING](https://nix-community.github.io/home-manager/#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [x] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/a51598236f23c89e59ee77eb8e0614358b0e896c/modules/programs/lesspipe.nix#L11).
  - [ ] Generate a news entry. See [News](https://nix-community.github.io/home-manager/index.xhtml#sec-news)
  - [ ] Basic tests added. See [Tests](https://nix-community.github.io/home-manager/index.xhtml#sec-tests)

- If this PR adds an exciting new feature or contains a breaking change.
  - [ ] Generate a news entry. See [News](https://nix-community.github.io/home-manager/index.xhtml#sec-news)

pinage404 commented:
I don't use it since a while

Could you remove me as a maintainer please ?

I may review later

s0racat commented:
Please **do not** merge this pull request.
I'll try to improve module and eliminate error

s0racat commented:
@pinage404 
I’ve finished fixing this module.
Please review it.

s0racat commented on modules/services/home-manager-auto-upgrade.nix:
```diff
@@ -57,23 +100,56 @@ in
+          Shell commands executed before `home-manager switch`.
+
+          - null: use legacy behavior (deprecated)
+          - []: run no pre-switch commands
```
Changed to use legacy behavior when preSwitchCommands is not set (=null).
I also find this implementation confusing

pinage404 commented on modules/services/home-manager-auto-upgrade.nix:
```diff
@@ -97,10 +173,17 @@ in
-        Unit.Description = "Home Manager upgrade";
+        Unit = {
+          Description = "Home Manager upgrade";
+          X-SwitchMethod = "keep-old";
```
Thanks !

s0racat commented on modules/services/home-manager-auto-upgrade.nix:
```diff
@@ -6,49 +6,90 @@
+    ];
+
     text =
       if cfg.useFlake then
```
I think we should deprecate useFlake and use flakeDir instead.

khaneliman commented on modules/misc/news/2026/01/2026-01-13_10-11-50.nix:
```diff
@@ -0,0 +1,12 @@
+{ pkgs, ... }:
+{
+  time = "2026-01-13T01:11:50+00:00";
+  condition = pkgs.stdenv.hostPlatform.isLinux;
```
Usually this is guarded for `config.home-manager.autoUpgrade.enable` so only people who use the module are alerted about the new functionality.

khaneliman commented on modules/services/home-manager-auto-upgrade.nix:
```diff
@@ -6,49 +6,90 @@
 in
 {
-  meta.maintainers = [ lib.maintainers.pinage404 ];
+  meta.maintainers = [ lib.hm.maintainers.soracat ];
```
Is it intended to drop @pinage404 ?

pinage404 replied:
Yes !

khaneliman replied:
Sorry, missed https://github.com/nix-community/home-manager/pull/8353#issuecomment-3658063282

khaneliman reviewed (changes requested):
Please fixup/squash commits to follow contribution guidelines.

s0racat committed:
maintainers: add soracat

s0racat committed:
home-manager-auto-upgrade: set maintainer to soracat

s0racat committed:
home-manager-auto-upgrade: add switch flags and pre-switch commands

Add a flags option for passing extra arguments to home-manager
switch and a preSwitchCommands option for running commands before the
switch.

Preserve the legacy flake update behavior behind a deprecation warning,
clean up the shell script, and cover the flake path in tests.

s0racat committed:
home-manager-auto-upgrade: clarify option documentation

Improve the option descriptions for enable, useFlake, flakeDir,
and the systemd timer format.

s0racat committed:
news: add home-manager-auto-upgrade options entry

Document the new flags and preSwitchCommands options for the
auto-upgrade service.

khaneliman committed:
home-manager-auto-upgrade: state-version gate preSwitchCommands

Migrate the preSwitchCommands default to
lib.hm.deprecations.mkStateVersionOptionDefault instead of using a
null sentinel.

Keep the legacy flake update behavior for older state versions and add
tests for the explicit, legacy, and current flake paths.

khaneliman commented:
Alright, organized the commits and migrated the default deprecation to use my new state version helper.

s0racat commented:
thank you for fixing!
